File: 121_BestTimetoBuyandSellStock.java
class Solution {
    public int maxProfit(int[] prices) {
        //dp[i] = max(dp[i-1], prices[i] - min(prices[0 : i]))
        int profit = 0;
        int cost = Integer.MAX_VALUE;
        for (int price : prices) {
            cost = Math.min(cost, price);
            profit = Math.max(profit, price - cost);
        }
        return profit;
    }
}

File: c++/121_BestTimetoBuyandSellStock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min_price = INT_MAX, res = 0;
        for (auto p : prices) {
            min_price = min(min_price, p);
            res = max(res, p - min_price);
        }
        return res;
    }
};